src/input/mod.rs

```diff
@@ -330,12 +332,63 @@ impl State {
             return;
         };
 
-        // Filter actions when the key is released or the session is locked.
-        if !pressed {
-            return;
-        }
+        let config = self.niri.config.clone();
+
+        if pressed {
+            self.handle_bind(bind.clone());
+
+            match self.niri.bind_repeat_timers.entry(bind.key) {
+                Entry::Vacant(entry) => {
+                    fn repeat_handler(state: &mut State, config: Rc<RefCell<Config>>, bind: Bind) {
+                        let repeat_after_millis = std::cmp::max(
+                            1000 / config.borrow().input.keyboard.repeat_rate as u64,
+                            100,
+                        );
+                        let repeat_rate_timer =
+                            Timer::from_duration(Duration::from_millis(repeat_after_millis));
+
+                        let bind_key = bind.key.clone();
+
+                        let token = state
+                            .niri
+                            .event_loop
+                            .insert_source(repeat_rate_timer, move |_, _, state| {
+                                state.handle_bind(bind.clone());
+                                repeat_handler(state, config.clone(), bind.clone());
+
+                                TimeoutAction::Drop
+                            })
+                            .unwrap();
 
-        self.handle_bind(bind);
+                        state.niri.bind_repeat_timers.insert(bind_key, token);
+                    }
+
+                    let repeat_delay_timer = Timer::from_duration(Duration::from_millis(
+                        config.borrow().input.keyboard.repeat_delay as u64,
+                    ));
+                    let token = self
+                        .niri
+                        .event_loop
+                        .insert_source(repeat_delay_timer, move |_, _, state| {
+                            state.handle_bind(bind.clone());
+                            repeat_handler(state, config.clone(), bind.clone());
+                            TimeoutAction::Drop
```

Comment:
Instead of re-creating the timer every time, you can return `TimeoutAction::Duration(repeat delay)` and it will re-fire for you. So the whole `repeat_handler()` is no longer needed.